Connection: Auto-disconnect on plugin deactivation

sergeymitr opened the pull request:
#### Changes proposed in this Pull Request:
This commit allows plugins using `jetpack-connection` to configure the package so it would automatically disconnect from WP.com when the plugin gets deactivated.

Whether soft or hard disconnect is used is also decided automatically, depending on other plugins utilizing the connection.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
Testing of this PR may not be easy, please ping me if you run into something unexpected.

*Single Site*
1. Install Jetpack and Jetpack Beta (checkout this branch), connect Jetpack.
2. Install Client-Example and checkout its branch `add/config-setup`.
3. Go to the `wp-content/plugins/client-example` directory, and create a symlink for the `vendor` dir:
```bash
ln -s /path/to/plugins/jetpack-dev/vendor
```
4. Connect Jetpack.
5. Deactivate Client Example, make sure Jetpack is still connected.
6. Activate Client Example, deactivate Jetpack. This will break the connection, so please connect back using Client Example.
7. Deactivate Client Example, activate it back.
8. Go to Client Example, confirm that it has been disconnected.

*Multisite Network*
1. Setup a multisite network with Jetpack and Jetpack Beta (checkout this branch).
2. Install Client Example and checkout its branch `add/config-setup`.
3. Go to the "Network -> Plugins" and make sure neither Jetpack nor Client Example are network-wide active.
4. Go to the `wp-content/plugins/client-example` directory, and create a symlink for the `vendor` dir:
```bash
ln -s /path/to/plugins/jetpack-dev/vendor
```
5. Create three additional sub-sites to get 4 subsites ready for testing.
6. Setup the sites in the following fashion:
**_When you activate Jetpack on sub-sites, make sure it's the `jetpack-dev` you're activating._**
_Site 1:_ Activate Jetpack and Client Example, connect Jetpack.
_Site 2:_ Activate Client-Example, connect it.
_Site 3:_ Activate Jetpack, connect it.
_Site 4:_ Don't activate anything, keep both Jetpack and Client Example disabled.
7. Go to Network Plugins and activate Client Example network-wide.
8. Go to Site 4 and connect it to WP.com (keep Jetpack disabled there).
9. Go to Network Plugins and deactivate Client Example.
10. Visit the sites:
_Site 1:_ Make sure Jetpack and Client Example are active and connected.
_Site 2:_ Confirm that Client Example is active and connected.
_Site 3:_ Confirm that Jetpack is active and connected.
_Site 4:_ Client Example should be disabled. Activate it and make sure it's not connected.

#### Proposed changelog entry for your changes:
TBD

sergeymitr committed:
Connection: Auto-disconnect on plugin deactivation.

This commit allows plugins using `jetpack-connection` to configure the package so it would automatically disconnect from WP.com.

Whether soft or hard disconnect is used is also decided automatically, depending on other plugins utilizing the connection.

dereksmart commented on packages/config/src/class-config.php:
```diff
@@ -71,6 +71,38 @@ public function ensure( $feature, array $options = array() ) {
+	 * @param array $common_options The common options to be passed into each feature's initialization method.
+	 * @param array $features       Features to be enabled and their options.
+	 */
+	public function ensure_bunch( array $common_options, array $features ) {
```
I know this is still in progress, but I'm excited! 

Do we need a new method? Can we not refactor the ->ensure() method to accept multiple features?

sergeymitr replied:
Yes, we can modify `ensure()` and check if the first argument is an array or a string.
I think it would make the code more complex, and having a separate passthrough method is easier to understand for the package users.

It's syntax is already complex (see the phpdoc example), and trying to merge it into the existing `ensure()` argument format will make it even even more complicated and error-prone.

dereksmart replied:
I guess it's a question of what is more important to us: Code complexity or usability/maintenance complexity? 

I feel like it's not ideal to have multiple methods that can `ensure` a feature/package that we need to maintain. This means double the tests, double to back-compat considerations, and double the entry points and syntaxes for setting up a plugin that we need to consider. 

I'm open to discuss further though, wonder what the team thinks.

sergeymitr replied:
> This means double the tests

We will still need separate tests for `ensure()` to check its behavior in the regular and multi-option mode.
And since `ensure_bunch()` is generally a passthrough method, we won't need to duplicate all the same tests we write for `ensure()`.
There are no tests for the `Config` package at the moment though ¯\\\_(ツ)\_/¯

I'll bring it up with the team to discuss.

dereksmart replied:
ok right, to me this is the biggest consideration: `double the entry points`. But worth a discussion!

sergeymitr committed:
Connection Auto-Disconnect: Minor adjustments.

- Renamed `ensure_bunch()` method to `config()`
- Changed the `config()` method arguments order

sergeymitr committed:
Connection: Auto-disconnect on multisite networks.

Multisite networks require the connection to be cleaned on site-by-site basis. This commit introduces the improved cleanup, when sites with no other connected plugins get cleaned, and the others will keep their connection intact.

gravityrail commented on packages/config/src/class-config.php:
```diff
@@ -208,6 +240,20 @@ protected function ensure_options_connection() {
+			$connection = new Manager( $slug );
+
+			if ( ! empty( $options['deactivate_disconnect'] ) ) {
+				if ( empty( $options['plugin_file'] ) ) {
```
You can also grab caller information from the PHP runtime and use that to figure out the slug of the calling plugin

gravityrail commented on packages/config/src/class-config.php:
```diff
@@ -71,6 +71,38 @@ public function ensure( $feature, array $options = array() ) {
+	 *
+	 * Usage example:
+	 * <code>
+	 * $config->setup(
```
This feels like a good step in the right direction, but I still think it could be simpler :) (at least for the most common cases)

I would really like things, as much as possible, to have sane defaults and a very very very lightweight initial implementation. Passing in big hashes is a recipe for complex and brittle integrations and it makes it hard to change strings later.

Ideally, in Boost, we would do something like this (pseudocode):

```php
use Automattic\Jetpack\Connection;

static function init() {
  $conn = Connection::init( // invoke the package class directly, rather than via some $config class, which encourages unnecessarily tight coupling between packages
    // $slug - a plugin identifier, defaults to this plugin's slug
    // $name - the plugin name, defaults to this plugin's translated name
    // $file - defaults to this plugin's main file
  );
}
```

Initializing the connection does the following:
- mounts API endpoints for reading and managing the connection
- emits connection state into the page for consumption by UI classes (see below)
- does any necessary plugin registration into a central config, if that's necessary for refcounting the API consumers

And in our UI code:

```jsx
import { ConnectButton, UserMenu, isConnected } from '@automattic/jetpack';

// ...

render() {
  return <>
     { isConnected() && <UserMenu /> }
     // and later
     { ! isConnected() && <ConnectButton /> }
  </>;
}
```

Similarly, if we want to initialize the Sync subsystem, we may invoke:

```php
use Automattic\Jetpack\Sync;
use Automattic\Jetpack\Connection;

static function init() {
  $conn = Connection::init();
  $sync = Sync::init( $conn );
  $sync->register_content_type( 'foo' );
  // or whatever you need to do with Sync
}
```

Now, because Sync has an explicit, initializer-level dependency on Connection (I am assuming here that Sync has a private constructor) it can operate like a "idempotent contextual singleton", giving each invoker a *connection-specific instance* of the Sync class so that operations can be namespaced, checked, etc against the Jetpack connection state of this specific plugin.

Now, let's look at how this would change some existing code, e.g. in the Jetpack plugin:

```php
public function configure() {
		$config = new Config();

		foreach (
			array(
				'sync',
			)
			as $feature
		) {
			$config->ensure( $feature );
		}

		$config->ensure(
			'connection',
			array(
				'slug' => 'jetpack',
				'name' => 'Jetpack',
			)
		);

		if ( is_admin() ) {
			$config->ensure( 'jitm' );
		}

		if ( ! $this->connection_manager ) {
			$this->connection_manager = new Connection_Manager( 'jetpack' );
		}

		/*
		 * Load things that should only be in Network Admin.
		 *
		 * For now blow away everything else until a more full
		 * understanding of what is needed at the network level is
		 * available
		 */
		if ( is_multisite() ) {
			$network = Jetpack_Network::init();
			$network->set_connection( $this->connection_manager );
		}
}
```

This would become:

```php
public function configure() {
		$this->connection= Connection::init();
		$this->sync = Sync::init( $connection );
		$this->connection = JITM::init( $connection );
		// none of the multisite stuff needs to be initialized because it's automatically detected and invoked by the Connection class, but of course you could also initialize it here
}
```

You could also include an additional singleton "get" function which would allow fetching a plugin-specific instance of a class. Effectively, if invoked from Plugin code, it would use the current plugin slug to find and return the correct instance of Sync or whatever for the plugin, and would throw an appropriate exception if the connection hasn't been initialized.

```php
// in some plugin
$jitm = JITM::get(); // throws "Plugin not connected" if it can't find the connection.
$jitm->render_ui(); // or something
```

Under the hood, you could create a centralized instance manager that allows you to manage references to classes, or a base class that contains the appropriate functionality.

For other classes that don't depend on the Connection, but perhaps do depend on each other, you can use a similar set of patterns, I would imagine.

Ultimately, this means we don't actually have (or need) a central entrypoint to configure the Jetpack DNA system as a whole. No "config" class that manages the initialization of all packages centrally. Instead, we have a bunch of coding conventions and attempt to use "zero config" techniques for initializing subsystems as much as possible. This makes this much easier for the plugin embedding the packages.

So, in the case of plugin deactivation, it would be enough just to call `Connection::init()` in most cases and the deactivation hooks will be taken care of for you!

sergeymitr replied:
Hi @gravityrail.
Thanks for such an extensive review of the approach.

We've been discussing it with the team over the last few days, and are planning provide a proper response later this week.

gravityrail reviewed (changes requested):
Left some long comments... hope they're useful!

sergeymitr commented:
We'll revisit this later.
